File: packages/sqlite_repo/src/sqliteSearcher.ts
import { Searcher, Envelope } from '@meshobj/common';
import { Auth } from '@meshobj/auth';
import { DTOFactory } from '@meshobj/graphlette';
import Handlebars from 'handlebars';
import { Database } from 'sqlite';

export class SQLiteSearcher implements Searcher {
    private db: Database;
    private table: string;
    private authorizer: Auth;
    private dtoFactory: DTOFactory;

    private singletonTemplate = Handlebars.compile(`
        SELECT *
        FROM {{_name}}
        WHERE {{{filters}}}
         AND created_at <= {{_created_at}}
        ORDER BY created_at DESC
        LIMIT 1
    `);

    private vectorTemplate = Handlebars.compile(
        `WITH latest AS (
        SELECT
            id,
            MAX(created_at) AS max_created_at
        FROM {{_name}}
        WHERE {{{filters}}}
            AND created_at <= {{_created_at}}
            AND deleted = 0
        GROUP BY id
        )
        SELECT t1.*
        FROM {{_name}} t1
        JOIN latest t2
        ON t1.id = t2.id
        AND t1.created_at = t2.max_created_at
        WHERE t1.deleted = 0`,
    );

    constructor(db: Database, table: string, dtoFactory: DTOFactory, authorizer: Auth) {
        this.db = db;
        this.table = table;
        this.dtoFactory = dtoFactory;
        this.authorizer = authorizer;
    }

    processQueryTemplate(parameters: Record<string, any>, queryTemplate: Handlebars.TemplateDelegate): string {
        return queryTemplate(parameters);
    }

    async find(
        queryTemplate: Handlebars.TemplateDelegate,
        args: Record<string, any>,
        creds: string[] = [],
        timestamp: number = Date.now(),
    ): Promise<Record<string, any>> {
        args._created_at = timestamp;
        args._name = this.table;
        args.filters = this.processQueryTemplate(args, queryTemplate);
        let sql = this.processQueryTemplate(args, this.singletonTemplate);

        const result = await this.db.get(sql, []);

        if (result && result.payload) {
            result.payload = JSON.parse(result.payload);
            result.authorized_tokens = JSON.parse(result.authorized_tokens);

            if (await this.authorizer.isAuthorized(creds, result)) {
                result.payload.id = result.id;
                return this.dtoFactory.fillOne(result.payload, timestamp);
            }
        }

        return {};
    }

    async findAll(
        queryTemplate: Handlebars.TemplateDelegate,
        args: Record<string, any>,
        creds: string[] = [],
        timestamp: number = Date.now(),
    ): Promise<Record<string, any>[]> {
        args._created_at = timestamp;
        args._name = this.table;
        args.filters = this.processQueryTemplate(args, queryTemplate);
        let sql = this.processQueryTemplate(args, this.vectorTemplate);

        const rows = await this.db.all(sql, []);
        const envelopes: Envelope[] = rows.map((i) => {
            i.payload = JSON.parse(i.payload);
            i.authorized_tokens = JSON.parse(i.authorized_tokens);
            return i;
        });

        const authorizedResults = envelopes.filter((row) => this.authorizer.isAuthorized(creds, row));

        return authorizedResults.map((row) => {
            const result = row as Envelope;
            result.payload.id = result.id;
            return result.payload;
        });
    }

    ready = async (): Promise<boolean> => {
        try {
            await this.db.get('SELECT 1');
            return true;
        } catch {
            return false;
        }
    };
}

File: java/meshql/repos/sqlite/src/main/java/com/meshql/repositories/sqlite/SQLiteSearcher.java
package com.meshql.repositories.sqlite;

import com.meshql.core.Auth;
import com.meshql.repositories.rdbms.RDBMSSearcher;

import javax.sql.DataSource;

public class SQLiteSearcher extends RDBMSSearcher {
    private static final String SINGLETON_QUERY_TEMPLATE = "SELECT e.*, " +
            "    (SELECT json_group_array(token) FROM " +
            "        (SELECT token FROM %s_authtokens " +
            "         WHERE envelope_id = e.id AND envelope_created_at = e.created_at " +
            "         ORDER BY token_order)) AS authorized_tokens " +
            "FROM %s e " +
            "WHERE %s " +
            "  AND e.created_at <= ? " +
            "  AND e.deleted = 0 " +
            "ORDER BY e.created_at DESC " +
            "LIMIT 1";

    /**
     * SQL template for finding multiple records.
     */
    private static final String VECTOR_QUERY_TEMPLATE = "WITH latest_versions AS (" +
            "    SELECT DISTINCT id, MAX(created_at) AS max_created_at " +
            "    FROM %s " +
            "    WHERE %s " +
            "      AND created_at <= ? " +
            "      AND deleted = 0 " +
            "    GROUP BY id" +
            ") " +
            "SELECT e.*, " +
            "    (SELECT json_group_array(token) FROM " +
            "        (SELECT token FROM %s_authtokens " +
            "         WHERE envelope_id = e.id AND envelope_created_at = e.created_at " +
            "         ORDER BY token_order)) AS authorized_tokens " +
            "FROM %s e " +
            "JOIN latest_versions lv ON e.id = lv.id AND e.created_at = lv.max_created_at " +
            "WHERE e.deleted = 0";

    public SQLiteSearcher(DataSource dataSource, String tableName, Auth authorizer) {
        super(SINGLETON_QUERY_TEMPLATE, VECTOR_QUERY_TEMPLATE, dataSource, tableName, authorizer);
    }
} 